l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,742 @@
+    /// are controllable through the Mailbox, since the Mailbox applies address aliasing to the from address for the
+    /// L2 tx if the L1 msg.sender is a contract. Without address aliasing for L1 contracts as refund recipients they
+    /// would not be able to make proper L2 tx requests through the Mailbox to use or withdraw the funds from L2, and
+    /// the funds would be lost.
+    /// @return l2TxHash The L2 transaction hash of deposit finalization
+    function depositLegacyErc20Bridge(
+        address _msgSender,
+        address _l2Receiver,
+        address _l1Token,
+        uint256 _mintValue,
+        uint256 _amount,
+        uint256 _l2TxGasLimit,
+        uint256 _l2TxGasPerPubdataByte,
+        address _refundRecipient
+    ) public payable override onlyLegacyBridge nonReentrant returns (bytes32 l2TxHash) {
+        require(l2BridgeAddress[ERA_CHAIN_ID] != address(0), "EB b. n dep");
+        uint256 l2Value;
+        {
+            bool ethIsBaseToken = (bridgehub.baseToken(ERA_CHAIN_ID) == ETH_TOKEN_ADDRESS);
+            require(ethIsBaseToken, "EB d.it n E chain");
+            if (_l1Token == l1WethAddress) {
+                require(msg.value + _amount == _mintValue, "EB wrong ETH sent weth d.it");
+                l2Value = _amount;
+                // we don't increase chainBalance in this case since we do it in bridgehubDepositBaseToken
+            } else {
+                require(_mintValue == msg.value, "EB w mintV");
+                l2Value = 0;
+
+                if (!hyperbridgingEnabled[ERA_CHAIN_ID]) {
+                    chainBalance[ERA_CHAIN_ID][_l1Token] += _amount;
+                }
+            }
+
+            /// we don't deposit the funds here, as we did that in the legacy bridge
+        }
+        bytes memory l2TxCalldata = _getDepositL2Calldata(ERA_CHAIN_ID, _msgSender, _l2Receiver, _l1Token, _amount);
+
+        {
+            // If the refund recipient is not specified, the refund will be sent to the sender of the transaction.
+            // Otherwise, the refund will be sent to the specified address.
+            // If the recipient is a contract on L1, the address alias will be applied.
+            address refundRecipient = _refundRecipient;
+            if (_refundRecipient == address(0)) {
+                refundRecipient = _msgSender != tx.origin
+                    ? AddressAliasHelper.applyL1ToL2Alias(_msgSender)
+                    : _msgSender;
+            }
+
+            l2TxHash = _depositSendTx(
+                ERA_CHAIN_ID,
+                _mintValue,
+                l2Value,
+                l2TxCalldata,
+                _l2TxGasLimit,
+                _l2TxGasPerPubdataByte,
+                refundRecipient
+            );
+        }
+
+        // for weth we don't save the depositHappened, since funds are sent to refundRecipient on L2 if the tx fails
+        if (_l1Token != l1WethAddress) {
```

Comment:
Can't we still claim failed deposit through the L1ERC20Bridge?

Reply:
No, because we need both `depositHappened` in the sharedBridge and  `depositAmount` in the L1ERC20Bridge for these types of withdrawals. This is 2x the storage writes, (I could remove that if needed, it would make it a bit more complicated though)